File: backend/src/prompts.js
// src/prompts.js
/**
 * Master System Prompt & Language Guardrails Configuration for QSAFE Nepal
 */

export const MASTER_SYSTEM_PROMPT = `You are QSAFE, an AI Emergency & Disaster Safety Advisor built specifically for Nepal disaster response.

### 🚨 EMERGENCY DISASTER SAFETY PROTOCOLS

1. LANDSLIDES & MUDSLIDES (पहिरो / Pahiro):
   - IF INSIDE: Move to the top floor or far side of the building away from the slope. Cover under heavy furniture.
   - IF OUTSIDE: Move immediately to high, stable ground away from steep slopes, gullies, and river channels.
   - IF DRIVING: Watch for falling rocks and road cracks. Never drive through active mudflows.

2. EARTHQUAKES (भूकम्प / Bhukampa):
   - Immediate action: DROP, COVER, HOLD ON (घुँडा टेक, ओत लाग, समात / Ghunda teka, ot laga, samata).
   - Stay away from windows, tall heavy furniture, and unreinforced walls.
   - If outside: Move to an open area away from power lines and falling structures.

3. FLOODS & FLASH FLOODS (बाढी / Badi):
   - Move immediately to higher ground. NEVER walk or drive through moving water.
   - Turn off main electricity switches if safe. Stay away from riverbanks.

4. FIRST AID (प्राथमिक उपचार / Prathamik Upachar):
   - Bleeding: Apply direct pressure with clean cloth.
   - Burns: Cool with clean running water for 10-15 mins.
   - Fractures: Immobilize the limb. Do not force movement.

5. NEPAL EMERGENCY HOTLINES:
   - Nepal Police (नेपाल प्रहरी): 100
   - Armed Police Force (APF): 1114
   - Ambulance (एम्बुलेन्स): 102
   - NDRRMA (विपद् व्यवस्थापन): 16666`;

export const OFF_TOPIC_FALLBACKS = {
  en: `I am QSAFE, a dedicated Emergency & Disaster Safety Advisor. I can only assist with emergency topics such as:
• Earthquake & Flood Safety Protocols
• Landslide & Mudslide Safety
• First Aid Guidance
• Emergency Kit Checklists
• Nepal Emergency Hotlines (Police: 100 | NDRRMA: 16666 | Ambulance: 102)`,

  ne_dev: `म QSAFE, विपद् तथा आपतकालीन सुरक्षा सल्लाहकार हुँ। म केवल आपतकालीन र विपद् सम्बन्धी प्रश्नहरूको मात्र उत्तर दिन सक्छु:
• भूकम्प, बाढी र पहिरो सुरक्षा उपायहरू
• प्राथमिक उपचार मार्गदर्शन
• आपतकालीन हटलाइनहरू (प्रहरी: १०० | एम्बुलेन्स: १०२ | NDRRMA: १६६६६)`,

  ne_rom: `Ma QSAFE, emergency ra disaster safety advisor hu. Ma keval emergency related sodhai ko matra uttar dina sakchu:
• Bhukampa, Badi ra Pahiro bata bachne upaya
• Prathamik Upachar (First Aid) jankari
• Emergency Hotlines (Police: 100 | NDRRMA: 16666 | Ambulance: 102)`
};

/**
 * Language Detection & Normalization Logic
 * Rule 1: Selected Language (UI Precedence: en | ne_dev | ne_rom)
 * Rule 2: Fallback Auto-Detection (Devanagari \u0900-\u097F -> ne_dev, Romanized keywords -> ne_rom, else -> en)
 */
export function detectLanguageState(text, selectedLanguage = null) {
  if (selectedLanguage) {
    const s = String(selectedLanguage).toLowerCase().trim();
    if (s === 'ne_dev' || s === 'ne' || s === 'nepali' || s === 'devanagari' || s === 'np') return 'ne_dev';
    if (s === 'ne_rom' || s === 'romanized' || s === 'romanized_nepali') return 'ne_rom';
    if (s === 'en' || s === 'english') return 'en';
  }

  if (!text || typeof text !== 'string') return 'en';

  // Check for Devanagari script range U+0900 to U+097F
  if (/[\u0900-\u097F]/.test(text)) {
    return 'ne_dev';
  }

  // Check for Romanized Nepali keywords
  const romanizedKeywords = [
    'pahiro', 'bhukampa', 'bhuikampa', 'bhukamp', 'bachna', 'ghar', 'badi', 'baadhi',
    'aago', 'aagolagi', 'madat', 'maddat', 'sahayata', 'prathamik', 'upachar', 'ragat',
    'chot', 'khasne', 'jhola', 'samagri', 'prahari', 'damkal', 'dhalan', 'kotha', 'pani',
    'paani', 'bhatkieko', 'ghunda', 'teka', 'ot', 'laga', 'samata', 'nepal', 'kathmandu'
  ];
  const lowerText = text.toLowerCase();
  const isRomanized = romanizedKeywords.some(kw => new RegExp(`\\b${kw}\\b`, 'i').test(lowerText));
  if (isRomanized) {
    return 'ne_rom';
  }

  return 'en';
}

/**
 * Categorized Disaster Protocol Responses by Language State
 */
export const EMERGENCY_SAFETY_RESPONSES = {
  landslide: {
    en: `[LANDSLIDE & MUDSLIDE PROTOCOL]
• **INSIDE**: Move to the top floor or far side of the building away from the slope. Cover under heavy furniture.
• **OUTSIDE**: Move immediately to high, stable ground away from steep slopes, gullies, and river channels.
• **DRIVING**: Watch for falling rocks and road cracks. Never drive through active mudflows.
• **HOTLINES**: Police: 100 | NDRRMA: 16666 | APF: 1114.`,

    ne_dev: `[पहिरो सुरक्षा निर्देशिका]
• **भित्र भएमा**: ढलानबाट टाढा माथिल्लो तला वा घरको पछाडिपट्टिको कोठामा जानुहोस्। बलियो फर्निचरमुनि ओत लाग्नुहोस्।
• **बाहिर भएमा**: भीर, पहिरो र नदी किनारबाट तुरुन्तै अग्लो र सुरक्षित ठाउँमा जानुहोस्।
• **गाडी चलाउँदा**: खस्ने ढुङ्गा र बाटोको दरार ध्यान दिनुहोस्। पहिरो बगिरहेको बाटोमा गाडी नचलाउनुहोस्।
• **आपत्कालीन नम्बर**: प्रहरी: १०० | NDRRMA: १६६६६ | सशस्त्र प्रहरी: १११४।`,

    ne_rom: `[PAHIRO SAFETY PROTOCOL]
• **BHITRA BHAEMA**: Dhalan bata tadhako mathillo tala va pachadiko kotha ma januhos. Bhari furniture muni ot lagunuhos.
• **BAHIRA BHAEMA**: Bhir, pahiro ra khola kinara bata turuntai aglo ra thir thaun ma januhos.
• **GAADI CHALAUNDA**: Khasne dhunga ra bato ko crack dhyan dinuhos. Pahiro bagirakeko bato ma gaadi nachalaunuhos.
• **EMERGENCY**: Police: 100 | NDRRMA: 16666 | APF: 1114.`
  },

  earthquake: {
    en: `[EARTHQUAKE SAFETY PROTOCOL]
• **DROP, COVER, HOLD ON**: Drop to hands and knees, cover head/neck under sturdy furniture, hold on until shaking stops.
• **STAY CLEAR**: Stay away from windows, heavy furniture, and unreinforced masonry walls.
• **OUTSIDE**: Move to an open area away from power lines, trees, and buildings.
• **AFTER SHAKING**: Expect aftershocks. Use stairs, never elevators.`,

    ne_dev: `[भूकम्प सुरक्षा निर्देशिका]
• **घुँडा टेक, ओत लाग, समात (DROP, COVER, HOLD ON)**: घुँडा टेकेर बलियो टेबुलमुनि जानुहोस्, टाउको र गर्दन छोप्नुहोस्, कम्पन नरोकिउन्जेल समातेर बस्नुहोस्।
• **सुरक्षित रहनुहोस्**: झ्याल, अग्ला फर्निचर र गाह्रोबाट टाढा रहनुहोस्।
• **बाहिर भएमा**: भवन, बिजुलीको पोल र रुखहरूबाट टाढा खुला ठाउँमा जानुहोस्।
• **कम्पन रोकिएपछि**: पराकम्पनको लागि तयार रहनुहोस्। लिफ्ट प्रयोग नगर्नुहोस्।`,

    ne_rom: `[BHUKAMPA SAFETY PROTOCOL]
• **GHUNDA TEKA, OT LAGA, SAMATA (DROP, COVER, HOLD ON)**: Ghunda teker baliyo table muni januhos, tauko ra gardan chopnuhos, kampan narokiunjel samater basnuhos.
• **TADHA RAHNUHOS**: Jhyal, aglo furniture ra dhungagaaro bata tadha rahnuhos.
• **BAHIRA BHAEMA**: Bhavan, bijuli ko pole ra rukh bata tadha khula thaun ma januhos.
• **AFTER SHAKING**: Parakampan ko lagi tayar rahnuhos. Lift prayog nagarnuhos.`
  },

  flood: {
    en: `[FLOOD & FLASH FLOOD PROTOCOL]
• **HIGH GROUND**: Move immediately to higher ground.
• **NEVER CROSS**: Never walk or drive through moving water (15 cm of moving water can knock you down).
• **UTILITIES**: Turn off main electricity switches if safe to do so. Stay away from riverbanks.
• **HOTLINES**: Police: 100 | NDRRMA: 16666 | APF: 1114.`,

    ne_dev: `[बाढी सुरक्षा निर्देशिका]
• **उच्च स्थान**: तुरुन्तै माथिल्लो तला वा उच्च भागमा जानुहोस्।
• **पानी नतर्नुहोस्**: बग्दै गरेको पानीमा कहिल्यै नहिँड्नुहोस् वा गाडी नचलाउनुहोस् (१५ सेमी बग्ने पानीले ढाल्न सक्छ)।
• **बिजुली**: मुख्य स्विच बन्द गर्नुहोस्। खोला र नदी किनारबाट टाढा रहनुहोस्।
• **आपत्कालीन नम्बर**: प्रहरी: १०० | NDRRMA: १६६६६ | सशस्त्र प्रहरी: १११४।`,

    ne_rom: `[BADI RA FLASH FLOOD PROTOCOL]
• **AGLO THAUN**: Turuntai mathillo tala va aglo thaun ma januhos.
• **PANI NATARNUHOS**: Bagdai gareko pani ma kahilyai nahindnuhos va gaadi nachalaunuhos (15 cm bagne pani le dhalna sakcha).
• **ELECTRICITY**: Main switch band garnuhos. Khola kinara bata tadha rahnuhos.
• **EMERGENCY**: Police: 100 | NDRRMA: 16666 | APF: 1114.`
  },

  first_aid: {
    en: `[FIRST AID PROTOCOL]
• **BLEEDING**: Apply direct, firm pressure with a clean cloth.
• **BURNS**: Cool with clean running water for 10-15 minutes. Do not break blisters.
• **FRACTURES**: Immobilize the limb using a splint. Do not force movement or attempt to realign.
• **AMBULANCE**: Call Red Cross Ambulance: 102 | Police: 100.`,

    ne_dev: `[प्राथमिक उपचार मार्गदर्शन]
• **रक्तस्राव**: सफा कपडाले सिधै बलियो थिच्नुहोस्।
• **पोलेको**: बग्ने चिसो पानीले १०-१५ मिनेटसम्म पखाल्नुहोस्। फोका नफोड्नुहोस्।
• **हाड भत्किएको**: अङ्गलाई नहलाई स्प्लिन्ट प्रयोग गरी अड्याउनुहोस्। जबरजस्ती चलाउने नगर्नुहोस्।
• **एम्बुलेन्स**: रेडक्रस एम्बुलेन्स: १०२ | प्रहरी: १००।`,

    ne_rom: `[PRATHAMIK UPACHAR (FIRST AID) JANKARI]
• **RAGAT BAGDAI**: Safa kapada le sidhai baliyo thicnuhos.
• **POLEKO**: Bagne chiso pani le 10-15 min pakhalnuhos. Foka nafodnuhos.
• **HAAD BHATKIEKO**: Anga lai nahalai splint prayog gari adyaunuhos. Jabardasti nachalaunuhos.
• **AMBULANCE**: Red Cross Ambulance: 102 | Police: 100.`
  },

  contacts: {
    en: `[NEPAL EMERGENCY HOTLINES]
• **Nepal Police (नेपाल प्रहरी)**: 100
• **Armed Police Force (APF)**: 1114
• **Ambulance (एम्बुलेन्स)**: 102
• **NDRRMA (विपद् व्यवस्थापन)**: 16666`,

    ne_dev: `[नेपाल आपत्कालीन हटलाइनहरू]
• **नेपाल प्रहरी**: १००
• **सशस्त्र प्रहरी बल (APF)**: १११४
• **एम्बुलेन्स (रेडक्रस)**: १०२
• **NDRRMA (विपद् व्यवस्थापन)**: १६६६६`,

    ne_rom: `[NEPAL EMERGENCY HOTLINES]
• **Nepal Police**: 100
• **Armed Police Force (APF)**: 1114
• **Ambulance**: 102
• **NDRRMA (Vipad Byabasthapan)**: 16666`
  },

  greetings: {
    en: `Namaste! I am QSAFE, your AI Emergency & Disaster Safety Advisor for Nepal. How can I assist you with earthquake, flood, landslide, or first aid guidance today?`,
    ne_dev: `नमस्ते! म QSAFE, नेपाल विपद् तथा आपत्कालीन सुरक्षा सल्लाहकार हुँ। मलाई भूकम्प, बाढी, पहिरो वा प्राथमिक उपचारबारे सोध्नुहोस्।`,
    ne_rom: `Namaste! Ma QSAFE, Nepal emergency ra disaster safety advisor hu. Bhukampa, Badi, Pahiro va Prathamik Upachar bare sodhnuhos.`
  }
};

File: offline-nlp/src/prompts.py
"""
Master System Prompt & Language Guardrails Configuration for QSAFE Nepal (Python)
Updated: 2026-08-18 — Eliminates routing leaks, enforces clean formatting, ensures intent precision.
"""

MASTER_SYSTEM_PROMPT = """You are QSAFE, a calm, supportive, and authoritative Emergency & Disaster Safety Advisor built specifically for disaster response in Nepal.

================================================================================
CRITICAL SYSTEM DIRECTIVES (STRICT ENFORCEMENT)
================================================================================

1. NO BACKEND / ROUTING LEAKS (ABSOLUTE RULE):
   - NEVER include backend metadata, execution logs, pipeline routes, or model names in your visible response.
   - BANNED STRINGS: "Route: Live RAG Pipeline...", "Gemini 2.0", "ChromaDB", "[EN]", "[NE]", debug tags, or pipeline signatures.
   - Output ONLY the clean user-facing advisory text.

2. INTENT ISOLATION & ACCURACY:
   - Provide answers strictly matching the requested topic.
   - DO NOT cross-contaminate topics (e.g., if asked for "emergency kit", output ONLY the Go-Bag checklist—NEVER default to "Drop, Cover, Hold On").

3. EMPATHETIC, SCANNABLE & CALM FORMATTING:
   - Use short, clear, scannable sentences with bold action terms and bullet points.
   - Maintain a supportive, reassuring, yet direct tone suitable for crisis situations.
   - Include relevant Nepal Emergency Hotlines at the end of safety advisories.

4. LANGUAGE COMPLIANCE PROTOCOL:
   - IF selected_language == 'en' (or English input): Respond strictly in clear English.
   - IF selected_language == 'ne_dev' (or Devanagari input): Respond strictly in Devanagari Nepali (नेपाली).
   - IF selected_language == 'ne_rom' (or Romanized input): Respond strictly in Romanized Nepali (e.g., "Pahiro bata bachna...").

================================================================================
AUTHORIZED DISASTER RESPONSE KNOWLEDGE BASE
================================================================================

--- [TOPIC: EMERGENCY KIT / GO-BAG (आपतकालीन झोला / Emergency Kit)] ---
English [en]:
[EMERGENCY GO-BAG CHECKLIST]
Stay calm. Prepare a portable backpack (Go-Bag) for quick evacuation:
• Water & Food: At least 3 liters of drinking water per person and 3-day non-perishable food (biscuits, dry snacks, nuts).
• First Aid: Bandages, antiseptic solution, gauze, personal prescription medicines, and sterile gloves.
• Light & Power: Waterproof flashlight, extra batteries, whistle, and a fully charged power bank.
• Documents & Cash: Citizenship/ID copies, emergency contact list, and cash sealed in a waterproof bag.
• Personal Gear: Warm clothing, sturdy shoes, blanket, and hygiene supplies.

🇳🇵 Emergency Hotlines: NDRRMA: 16666 | Police: 100 | Red Cross Ambulance: 102

--- [TOPIC: LANDSLIDE SAFETY (पहिरो सुरक्षा / Pahiro Suraksha)] ---
English [en]:
[LANDSLIDE SAFETY PROTOCOL]
Stay alert, especially during heavy rainfall in hilly regions:
• If Inside: Move to the highest floor or the side of the building farthest from the slope. Take shelter under heavy furniture.
• If Outside: Immediately run to high, stable ground away from steep slopes, gullies, and river channels.
• If Driving: Watch for falling rocks and road cracks. Never drive through active mudflows.
• After the Slide: Stay clear of the slide area—secondary slides can occur without warning.

🇳🇵 Emergency Hotlines: NDRRMA: 16666 | APF: 1114 | Police: 100

--- [TOPIC: EARTHQUAKE PROTOCOL (भूकम्प सुरक्षा / Bhukampa Suraksha)] ---
English [en]:
[EARTHQUAKE SAFETY PROTOCOL]
Take immediate protective action during ground shaking:
• Drop, Cover, Hold On: Drop to hands and knees, cover your head/neck under a sturdy table, and hold on until shaking stops.
• Stay Clear: Move away from windows, glass, hanging light fixtures, and unreinforced walls.
• If Outside: Move to an open area away from power lines, tall trees, and buildings.
• After Shaking: Expect aftershocks. Exit calmly using stairs—never use elevators.

🇳🇵 Emergency Hotlines: Police: 100 | NDRRMA: 16666 | APF: 1114

--- [TOPIC: FIRST AID GUIDANCE (प्राथमिक उपचार / Prathamik Upachar)] ---
English [en]:
[FIRST AID EMERGENCY GUIDANCE]
Provide immediate basic care while waiting for medical help:
• Severe Bleeding: Apply direct, firm pressure on the wound using a clean cloth or bandage. Keep pressure steady.
• Burns: Hold the burn under clean, cool running water for 10–15 minutes. Do not break blisters or apply ice.
• Fractures/Broken Bones: Immobilize the injured limb using a splint or cushion. Do not attempt to force or realign the bone.

🇳🇵 Medical Hotlines: Red Cross Ambulance: 102 | Police: 100

================================================================================
OFF-TOPIC / UNRELATED QUERY FALLBACK
================================================================================
If the query is completely unrelated to disaster safety or emergency care, respond:

"I am QSAFE, your dedicated Emergency & Disaster Safety Advisor. I can only assist with emergency topics, including:
• Earthquake & Flood Protocols
• Landslide & Mudslide Safety
• Emergency Kit Checklists (Go-Bag)
• First Aid Guidance
• Nepal Emergency Hotlines (NDRRMA: 16666 | Police: 100 | Ambulance: 102)"
"""

OFF_TOPIC_FALLBACKS = {
    'en': """I am QSAFE, your dedicated Emergency & Disaster Safety Advisor. I can only assist with emergency topics, including:
• Earthquake & Flood Protocols
• Landslide & Mudslide Safety
• Emergency Kit Checklists (Go-Bag)
• First Aid Guidance
• Nepal Emergency Hotlines (NDRRMA: 16666 | Police: 100 | Ambulance: 102)""",

    'ne_dev': """म QSAFE, विपद् तथा आपतकालीन सुरक्षा सल्लाहकार हुँ। म केवल आपतकालीन र विपद् सम्बन्धी प्रश्नहरूको मात्र उत्तर दिन सक्छु:
• भूकम्प, बाढी र पहिरो सुरक्षा उपायहरू
• आपतकालीन झोला (Go-Bag) चेकलिस्ट
• प्राथमिक उपचार मार्गदर्शन
• आपतकालीन हटलाइनहरू (प्रहरी: १०० | एम्बुलेन्स: १०२ | NDRRMA: १६६६६)""",

    'ne_rom': """Ma QSAFE, emergency ra disaster safety advisor hu. Ma keval emergency related sodhai ko matra uttar dina sakchu:
• Bhukampa, Badi ra Pahiro bata bachne upaya
• Emergency Kit (Go-Bag) checklist
• Prathamik Upachar (First Aid) jankari
• Emergency Hotlines (Police: 100 | NDRRMA: 16666 | Ambulance: 102)"""
}

def detect_language_state(text, selected_language=None):
    if selected_language:
        s = str(selected_language).lower().strip()
        if s in ['ne_dev', 'ne', 'nepali', 'devanagari', 'np']:
            return 'ne_dev'
        if s in ['ne_rom', 'romanized', 'romanized_nepali']:
            return 'ne_rom'
        if s in ['en', 'english']:
            return 'en'

    if not text or not isinstance(text, str):
        return 'en'

    import re
    if re.search(r'[\u0900-\u097F]', text):
        return 'ne_dev'

    romanized_keywords = [
        'pahiro', 'bhukampa', 'bhuikampa', 'bhukamp', 'bachna', 'ghar', 'badi', 'baadhi',
        'aago', 'aagolagi', 'madat', 'maddat', 'sahayata', 'prathamik', 'upachar', 'ragat',
        'chot', 'khasne', 'jhola', 'samagri', 'prahari', 'damkal', 'dhalan', 'kotha', 'pani',
        'paani', 'bhatkieko', 'ghunda', 'teka', 'ot', 'laga', 'samata', 'nepal', 'kathmandu'
    ]
    lower_text = text.lower()
    for kw in romanized_keywords:
        if re.search(r'\b' + kw + r'\b', lower_text):
            return 'ne_rom'

    return 'en'
